### PR TITLE
synthesize vjp

### PR DESCRIPTION
**Pull request by marcrasi**
This PR synthesizes the VJP.

Importantly, I create the VJP immediately when the task is registered, so that PrimalGen will be able to reference the VJP. This required me to also create the empty primal and adjoint immediately when the task is registered, so that the VJP can reference the primal and adjoint. This means that we can no longer check if primal/adjoint needs to be synthesized by checking if `task.primal`/`task.adjoint` is `nullptr`, so I created a new piece of state `FunctionSynthesisState` to keep track of that.

The VJP synthesis itself, in `DifferentiationTask::createVJP` is a pretty straightforward imitation of the existing gradient synthesis code.

Miscellaneous changes required to make it all work:
* A lot of the tests differentiate functions with args/results that do not conform to `Differentiable`. This breaks `getAutoDiffAssociatedFunctionType()`, which `createVJP()` calls. So I added a quick workaround to make `getAutoDiffAssociatedFunctionType()` treat types as their own Tangent/Cotangent when they do not conform to `Differentiable`.
* `getAutoDiffAssociatedFunctionType()` needed a `resultIndex` to make some tests pass. I think we'll eventually want to allow multiple differentiated results? This single differentiated result makes it compatible with all the existing code, and we can add multiple results later.
* Rearranged the order of some sil functions in some tests because now they get generated in a different order.

I added on SIL test for the synthesized VJP. Once I make the PR that uses the VJP, I should be able to test it more thoroughly by actually executing it.

### CODE REVIEW

**Comment by rxwei** [comment]
> `getAutoDiffAssociatedFunctionType()` needed a `resultIndex` to make some tests pass. I think we'll eventually want to allow multiple differentiated results? This single differentiated result makes it compatible with all the existing code, and we can add multiple results later.

We'll support multiple results only at the SIL level, as the AST-level syntax is incapable of expressing result selection in a type-safe way. In today's SIL-level implementation, expecting a single result index is actually incorrect for general JVPs/VJPs. It doesn't have to be a single result -- it should be a set of result indices represented by a bit vector. Result selection determines the "vector" type in your SIL-level VJP function. For example, `(T, U) -> (V, W, R)` with result selection (1, 0, 1) will make the vector type be `(V.CotangentVector, R.CotangentVector)`. This also means that we will need a result-level indices representation and syntax, e.g. `(@nondiff T, U) -> (V, @nondiff W, R)`, only for SIL functions.

We won't be blocked on it for anything we need this month, though.

File: lib/SILOptimizer/Mandatory/TFDifferentiation.cpp
Comment (rxwei) [inline lib/SILOptimizer/Mandatory/TFDifferentiation.cpp]: Align subexpressions

**Review by rxwei — APPROVED** [review]

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

File: include/swift/AST/Types.h
Comment (marcrasi) [reply include/swift/AST/Types.h]: I will defer this change until later so that I can get this PR merged sooner. I do not think it will be much harder to make this change when we need it than to make this change now.

File: lib/AST/Type.cpp
Comment (marcrasi) [reply lib/AST/Type.cpp]: I will add a todo for the tuple thing.

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

File: include/swift/AST/Types.h
Comment (rxwei) [reply include/swift/AST/Types.h]: Ok. That's fine.

File: lib/SILOptimizer/Mandatory/TFDifferentiation.cpp
Comment (rxwei) [reply lib/SILOptimizer/Mandatory/TFDifferentiation.cpp]: We don't have the ability to do PrimalGen or AdjointGen for generic functions yet. How is the VJP synthesized?

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

File: lib/SILOptimizer/Mandatory/TFDifferentiation.cpp
Comment (marcrasi) [reply lib/SILOptimizer/Mandatory/TFDifferentiation.cpp]: Discussed in person.

For anyone reading: The next PR (https://github.com/apple/swift/pull/21062) can create the empty shells of generic primals and adjoints, but it still bails out when PrimalGen tries to start filling in the primal.

The next PR can also create a full generic VJP with body that calls primal and adjoint. So if the primal and adjoint for a generic function are manually specified, then a fully-working VJP gets created. This is just a convenience so that I can test that VJPs work on code that has custom primal/adjoint and no custom VJP. Soon this convenience will become irrelevant because we will forbid custom primal/adjoint and replace them with custom VJP.

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow linux gpu

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow linux gpu

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow macos

**Pull request merged**
